### DISCUSSION TITLE
How to send images by Dicom dataset

### DISCUSSION THREAD

**Discussion by foroughi9300** [fo-dicom/fo-dicom#1710] [General]
Mehdi Foroughi's profile photo
Mehdi Foroughi
11:27 AM (1 minute ago) 
to Fellow Oak DICOM
Hello,
I am reading a dicomdir file. Then I save all the datasets corresponding to each file. Then I create a DicomCStoreRequest for each dataset and add it to a list and the list of requests to a dicom client and send:

List<DicomCStoreRequest> cStoreRequestsList = new List<DicomCStoreRequest>();
for (int i = 0; i < dicomDatasetList.Count; i++)
{
        DicomDataset ds = dicomDatasetList[i];
        DicomCStoreRequest storeReq = new DicomCStoreRequest(ds) ;
        cStoreRequestsList.Add(storeReq);
}
        var client = new Dicom.Network.Client.DicomClient("127.0.0.1", 104, false, "MyAE", "CalledAe");
        await client.AddRequestsAsync(cStoreRequestsList.ToArray());
        await client.SendAsync();

but I get different kind of errors.
How can I send dicom datasets instead of sending dicom files?

**Comment by gofal** [comment]
I wonder what's the content of `dicomDatasetList`. You said you are reading a DICOMDIR file. You know, that a DICOMDIR only contains very minimal metadata. the most important tag is 0004,1500 Reference File Id. This is the filename of the full dicomimage relative to the directory where the DICOMDIR file is located. With that you can build up a filename and then have to create a DicomCStoreRequest from that filename, or you open the DicomFile and create a DicomCStoreRequest from that DicomFile

**Reply by foroughi9300** [reply]
Yes I think you are right... 
I am trying to scan a CD's contents and send it to somewhere else however it takes a lot of time to load the files first and then sending them whereas a software like Radiant DICOM viewer immediately starts showing and reading DICOM files. I wonder how can Radiant be so fast at reading and detecting DICOM files event when reading CDs that have no DICOMDIR.

**Reply by gofal** [reply]
By DICOM standard all CDs need to have a dicomdir file. It is not optional.
If there was an invalid CD without dicomdir then of course radiant will also be slow with reading the content.
